### PR TITLE
Codebridge: better resetting of status on level change

### PR DESCRIPTION
**Pull request by molly-moen**
This PR cleans up how we reset state/status when changing levels in Codebridge (specifically Python Lab, some features are cross-codebridge). It also includes some small related improvements to state/level switching. The level change notifier triggers both when switching levels and when a teacher switches to view a student's project on a level, which is what we want.

I used the approach of making a hook, `useLifecycleNotifier`, that registers a lifecycle notifier via a `useEffect`, and unregisters it when the `useEffect` returns. This approach is especially good for places where the state we are resetting is component-specific, and we have to do the resetting within the component. I'm less convinced when I'm resetting redux values (see `ControlButtons` where I reset some run state), but I wasn't sure of a good way to hook up a lifecycle notifier in the redux file itself. I'm open to suggestions on any improvements here!

I considered if we would want some sort of single source of truth location where we do all these resets. Again, it doesn't make sense to do that for component-level state, so I went with the approach of adding the hook where it made the most sense.

I made the following reset fixes:
- Clear the console
- Clear the run state
- Ensure any in-progress Python program is stopped
- Clear the version history status (viewing a previous project, just restored a project)

I also made the following small improvements:
- I made the stop and restart python method smarter; it only kills pyodide if a program is in progress.
- I updated the run/validate buttons such that if we are validating, run gets disabled, and vice versa (previously they stayed enabled when the other was running)

## Links
- jira ticket: [CT-790](https://codedotorg.atlassian.net/browse/CT-790)

## Testing story
Tested locally.

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/lab2/hooks/useLifecycleNotifier.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import {useEffect} from 'react';
```
Comment (molly-moen) [inline apps/src/lab2/hooks/useLifecycleNotifier.ts:1]: @sanchitmalhotra126 tagged you on this PR for thoughts/fyi that I'm adding this hook.

Comment (sanchitmalhotra126) [reply apps/src/lab2/hooks/useLifecycleNotifier.ts:1]: Nice, this seems really handy!

**Review by sanchitmalhotra126 — APPROVED** [review]
Lab2 changes LGTM!

**Review by thomasoniii — APPROVED** [review]
I ❤️ custom hooks.

lgtm 🚀

**Pull request merged**